Mathlib/Tactic/LibrarySearch.lean

```diff
@@ -98,22 +129,22 @@ open Lean.Parser.Tactic
 
 -- TODO: implement the additional options for `library_search` from Lean 3,
 -- in particular including additional lemmas
--- with `library_search [X, Y, Z]` or `library_search with attr`,
--- or requiring that a particular hypothesis is used in the solution, with `library_search using h`.
+-- with `library_search [X, Y, Z]` or `library_search with attr`.
 syntax (name := librarySearch') "library_search" (config)? (simpArgs)?
-  (" using " (colGt binderIdent)+)? : tactic
+  (" using " (colGt term),+)? : tactic
 syntax (name := librarySearch!) "library_search!" (config)? (simpArgs)?
-  (" using " (colGt binderIdent)+)? : tactic
+  (" using " (colGt term),+)? : tactic
 
 -- For now we only implement the basic functionality.
 -- The full syntax is recognized, but will produce a "Tactic has not been implemented" error.
 
 open Elab.Tactic Elab Tactic in
-elab_rules : tactic | `(tactic| library_search%$tk) => do
+elab_rules : tactic | `(tactic| library_search%$tk $[using $[$required:term],*]?) => do
   let mvar ← getMainGoal
   let (_, goal) ← (← getMainGoal).intros
   goal.withContext do
-    if let some suggestions ← librarySearch goal (← librarySearchLemmas.get) then
+    let required ← (required.getD {}).toList.mapM (elabTerm ·.raw none)
```

Comment:
```suggestion
    let required := (← getFVarIds (required.getD #[])).toList.map .fvar
```

Reply:
This doesn't work, as although we're still formally in `TacticM`, we've started doing manual goal handling here, so `getFVarIds` operates with the wrong context.

```
let required := (← (required.getD #[]).mapM getFVarId).toList.map .fvar
```
works fine.

But I wonder whether this declaration is stylistically bad in any case, and we should put a `liftMetaTactic` in explicitly, so that it is clear that we shouldn't be using `TacticM` internally. Happy for advice on this one, @gebner.

Reply:
I guess `addExactSuggestion` needs to be higher than `MetaM` anyway, so it's just a matter of being careful here.